roslibrust_codegen/src/utils.rs

```diff
@@ -129,3 +149,58 @@ fn message_files_from_path(path: &Path, ext: &str) -> io::Result<Vec<PathBuf>> {
 
     Ok(msg_files)
 }
+
+fn determine_ros_version(
+    path: impl AsRef<Path> + std::fmt::Debug,
+) -> Result<RosVersion, std::io::Error> {
+    use std::fs::File;
+    use std::io::BufReader;
+    use xml::reader::{EventReader, ParserConfig, XmlEvent};
+    const TAG: &'static str = "buildtool_depend";
+
+    let file = File::open(&path)?;
+    let reader = BufReader::new(file);
+    let parser = EventReader::new_with_config(
+        reader,
+        ParserConfig {
+            trim_whitespace: true,
+            ignore_comments: true,
+            ..Default::default()
+        },
+    );
+
+    let mut in_tag = false;
+    for e in parser {
```

Comment:
I remember doing this when trying to detect metapackages and I truly hate this API. It looks useful for streamed data, but it's such a pain to have to do state tracking manually.

Reply:
Yeah I'm defo not impressed, but it was pretty quick to get it to do what I wanted...